services/src/pro/permissions/postgres_permissiondb.rs

```diff
@@ -0,0 +1,198 @@
+use async_trait::async_trait;
+use snafu::ensure;
+use tokio_postgres::{
+    tls::{MakeTlsConnect, TlsConnect},
+    Socket,
+};
+use uuid::Uuid;
+
+use crate::error::{self, Error, Result};
+use crate::pro::contexts::PostgresDb;
+
+use super::{Permission, PermissionDb, ResourceId, RoleId};
+
+// TODO: a postgres specific permission db trait that allows re-using connections and transactions
+
+trait ResourceTypeName {
+    fn resource_type_name(&self) -> &'static str;
+
+    fn uuid(&self) -> Result<Uuid>;
+}
+
+impl ResourceTypeName for ResourceId {
+    fn resource_type_name(&self) -> &'static str {
+        match self {
+            ResourceId::Layer(_) => "layer_id",
+            ResourceId::LayerCollection(_) => "layer_collection_id",
+            ResourceId::Project(_) => "project_id",
+            ResourceId::DatasetId(_) => "dataset_id",
+        }
+    }
+
+    fn uuid(&self) -> Result<Uuid> {
+        match self {
+            ResourceId::Layer(id) => Uuid::parse_str(&id.0).map_err(|_| Error::InvalidUuid),
+            ResourceId::LayerCollection(id) => {
+                Uuid::parse_str(&id.0).map_err(|_| Error::InvalidUuid)
+            }
+            ResourceId::Project(id) => Ok(id.0),
+            ResourceId::DatasetId(id) => Ok(id.0),
+        }
+    }
+}
+
+#[async_trait]
+impl<Tls> PermissionDb for PostgresDb<Tls>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    async fn create_resource<R: Into<ResourceId> + Send + Sync>(&self, resource: R) -> Result<()> {
+        let resource: ResourceId = resource.into();
+
+        let conn = self.conn_pool.get().await?;
+
+        let stmt = conn
+            .prepare(&format!(
+                "
+            INSERT INTO permissions (role_id, permission, {resource_type})
+            VALUES ($1, $2, $3) ON CONFLICT DO NOTHING;",
```

Comment:
we do so much "?" here we might also report if there is a conflict. One can still ignore that error if not important? At least the user/admin could ge the feedback that the permission was already there? (same for all the others)

Reply:
yeah, it *might* make sense do this. But would we also throw an error if someone changes a Layer Name but the new value is the same is the old one? 

For me  this was not straight forward enough to put in the work to implement this. But if it's clear to you that we want this behavior, I can add it.

Reply:
good point. i think we should add a TODO and then talk about that? I think we can improve the feedback at a later point.

Reply:
I added a TODO here 00b3a35085a1cb2935b26841b8c19cd4246b1ae4